server/services/zwavejs-ui/lib/zwaveJSUI.setValue.js

```diff
@@ -49,57 +52,96 @@ function getNodeFeature(node, nodeFeatureId) {
   }
 
-  const nodeFeature = getNodeFeature(node, deviceFeature.external_id);
+  const zwaveJsNode = this.getZwaveJsDevice(node.external_id);
+  if (!zwaveJsNode) {
+    throw new BadParameters(`ZWaveJs-UI node not found: "${node.external_id}".`);
+  }
+
+  const nodeFeature = getNodeFeature(node, gladysFeature.external_id);
   if (!nodeFeature) {
-    throw new BadParameters(`ZWaveJs-UI feature not found: "${deviceFeature.external_id}".`);
+    throw new BadParameters(`ZWaveJs-UI feature not found: "${gladysFeature.external_id}".`);
   }
 
-  const command = getCommand(nodeFeature);
-  if (!command) {
-    // We do not manage this feature for writing
-    throw new BadParameters(`ZWaveJS-UI command not found: "${deviceFeature.external_id}"`);
+  const actionDescriptor = getAction(zwaveJsNode, nodeFeature);
+  if (!actionDescriptor) {
+    // We do not manage this feature for setValue
+    throw new BadParameters(`ZWaveJS-UI action not found: "${gladysFeature.external_id}"`);
   }
 
-  const commandArgs = command.getArgs(value, nodeFeature);
-  if (commandArgs === null) {
-    throw new BadParameters(`ZWaveJS-UI command value not supported: "${value}"`);
+  const nodeContext = { node, nodeFeature, zwaveJsNode, gladysDevice, gladysFeature };
+  const action = actionDescriptor(value, nodeContext);
+  if (action.isCommand) {
+    // API sendCommand
+    // https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=sendcommand
+    const mqttPayload = {
+      args: [
+        {
+          nodeId: nodeFeature.node_id,
+          commandClass: nodeFeature.command_class,
+          endpoint: nodeFeature.endpoint,
+        },
+        action.name,
+        action.value,
+      ],
+    };
+    this.publish('zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set', JSON.stringify(mqttPayload));
+  } else {
+    // API writeValue
+    // https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=writevalue
+    const mqttPayload = {
+      args: [
+        {
+          nodeId: nodeFeature.node_id,
+          commandClass: nodeFeature.command_class,
+          endpoint: nodeFeature.endpoint,
+          property: action.name,
+        },
+        action.value,
+      ],
+    };
+    this.publish('zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/writeValue/set', JSON.stringify(mqttPayload));
   }
 
-  // https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=send-command
-  // https://zwave-js.github.io/zwave-js-ui/#/guide/mqtt?id=sendcommand
-  const mqttPayload = {
-    args: [
-      {
-        nodeId: nodeFeature.node_id,
-        commandClass: nodeFeature.command_class,
-        endpoint: nodeFeature.endpoint,
-      },
-      command.getName(nodeFeature),
-      commandArgs,
-    ],
-  };
-  this.publish('zwave/_CLIENTS/ZWAVE_GATEWAY-zwave-js-ui/api/sendCommand/set', JSON.stringify(mqttPayload));
+  if (action.stateUpdate) {
+    const promises = [];
+    action.stateUpdate.forEach((stateUpdate) => {
```

Comment:
Maybe let's replace this with Bluebird Promise.map with a concurrency?

Reply:
Moved to Promise.map from Bluebird